jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/EvictingCachingContentFactory.java

```diff
@@ -0,0 +1,165 @@
+ * Uses a validationTime parameter to check files when they are accessed and/or sweepDelay parameter for a periodic sweep
+ * for invalid entries in the cache.
+ * </p>
+ * <p>
+ * {@link HttpContent} validation checks are configured through the {@code validationTime} parameter in the constructor.
+ * If an {@link HttpContent} is found to be invalid it will be removed from the cache.
+ * A value of -1 means that the cached {@link HttpContent} will never be checked if it is still valid,
+ * a value of 0 means it is checked on every request,
+ * a positive value indicates the number of milliseconds of the minimum time between validation checks.
+ * </p>
+ * <p>
+ * This also remember a missed entry for the time set by {@code validationTime}ms. After this has
+ * elapsed the entry will be invalid and will be evicted from the cache at the next access.
+ * </p>
+ */
+public class EvictingCachingContentFactory extends CachingHttpContentFactory implements Runnable
+{
+    private final Scheduler _scheduler;
+    private final long _sweepDelay;
+    private final long _validationTime;
+    private final long _maxCacheIdleTime;
+
+    public EvictingCachingContentFactory(@Name("authority") HttpContent.Factory authority,
+                                         @Name("validationTime") long validationTime)
+    {
+        this(authority, validationTime, null, -1, -1);
+    }
+
+    public EvictingCachingContentFactory(@Name("authority") HttpContent.Factory authority,
+                                         @Name("validationTime") long validationTime,
+                                         @Name("scheduler") Scheduler scheduler,
+                                         @Name("sweepDelay") long sweepDelay,
+                                         @Name("maxCacheIdleTime") long maxCacheIdleTime)
+    {
+        super(authority);
+        _validationTime = validationTime;
+        _scheduler = scheduler;
+        _sweepDelay = sweepDelay;
+        _maxCacheIdleTime = maxCacheIdleTime;
+        if (scheduler != null && sweepDelay > 0)
+            schedule();
+    }
+
+    @Override
+    protected boolean isCacheable(HttpContent httpContent)
+    {
+        if (httpContent == null)
+            return (_validationTime != 0);
+        return super.isCacheable(httpContent);
+    }
+
+    private void schedule()
+    {
+        _scheduler.schedule(this, _sweepDelay, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void run()
+    {
+        ConcurrentMap<String, CachingHttpContent> cache = getCache();
+        for (Map.Entry<String, CachingHttpContent> entry : cache.entrySet())
+        {
+            CachingHttpContent value = entry.getValue();
+            if (_maxCacheIdleTime > 0 && NanoTime.since(value.getLastAccessedNanos()) > TimeUnit.MILLISECONDS.toNanos(_maxCacheIdleTime))
+                removeFromCache(value);
+            else if (!value.isValid())
+                removeFromCache(value);
+        }
+        schedule();
```

Comment:
maybe put in a try-finally... just in case.

Reply:
Done.